File: CodingTest/Java/Solution26.java
// https://school.programmers.co.kr/learn/courses/30/lessons/64065
package CodingTest.Java;

import java.util.Arrays;
import java.util.Comparator;

public class Solution26 {
    public int[] solution(String s) {
        String t = s.substring(2, s.length() - 2);
        String[] temp = t.split("\\},\\{");
        int[] answer = new int[temp.length];
        Arrays.sort(temp, new Comparator<String>() {
            @Override
            public int compare(String o1, String o2) {
                return o1.length() - o2.length();
            }
        });
        
        answer[0] = Integer.parseInt(temp[0]);

        for (int i = 1; i < temp.length; i++) {
            int sumOfTemp = Arrays.stream(temp[i].split(",")).mapToInt(Integer::parseInt).sum();
            for (int j = 0; j < i; j++) {
                sumOfTemp -= answer[j];
            }
            answer[i] = sumOfTemp;
        }

        return answer;
    }
}

File: CodingTest/Python/solution26.py
# https://school.programmers.co.kr/learn/courses/30/lessons/258705

def solution(n, tops):
	lst = [0 for i in range(len(tops)*2 + 2)] # 0 1 2 3 4 5 6 7 8 9
	dp = [1,1]
	for i in range(len(tops)):
		lst[2*i+2] = tops[i] # 0 1 2 3
	for i in range(2, len(lst)):
		if lst[i] == 1:
			dp.append((dp[i-1]*2 + dp[i-2])%10007)
		else:
			dp.append((dp[i-1] + dp[i-2])%10007)
	return dp[-1]%10007